.github/scripts/transform_dbt_test_results.py

```diff
@@ -0,0 +1,1082 @@
+            f"status={self.status!r}, "
+            f"num_tests={len(self.test_results)}, "
+            f"num_failing_rows={num_failing_rows})"
+        )
+
+    @classmethod
+    def from_dict(cls, category_dict: typing.Dict) -> "TestCategory":
+        """Deserialize a TestCategory object from a dictionary."""
+        return TestCategory(
+            category=category_dict["category"],
+            results=[
+                TestResult.from_dict(result_dict)
+                for result_dict in category_dict["test_results"]
+            ],
+        )
+
+    @property
+    def fieldnames(self) -> typing.List[str]:
+        """Get a list of fieldnames that encapsulates all of the fieldnames
+        for all of the rows of tests tracked by this group."""
+        fieldnames = []
+        for result in self.test_results:
+            for fieldname in result.fieldnames:
+                if fieldname not in fieldnames:
+                    fieldnames.append(fieldname)
+
+        # Remove any fixed fieldnames from the ordered list that are not
+        # present in this group
+        fixed_field_order = [
+            field for field in self._fixed_fieldnames if field in fieldnames
+        ]
+
+        # Reorder the fieldnames so that diagnostic fields are presented in the
+        # correct order
+        for field in reversed(fixed_field_order):
+            fieldnames.insert(0, fieldnames.pop(fieldnames.index(field)))
+
+        return fieldnames
+
+    @property
+    def rows(self) -> typing.List[typing.List]:
+        """Format the rows of tests tracked by this group, with
+        fieldname data excluded. The combination of this property and the
+        `fieldnames` property can be used to write to a csv.Writer or
+        to an openpyxl.Workbook sheet for the tests tracked by this group."""
+        fieldnames = self.fieldnames
+        return [
+            [row.get(fieldname) for fieldname in fieldnames]
+            for result in self.test_results
+            for row in result.failing_rows
+        ]
+
+    @property
+    def status(self) -> Status:
+        """Return an aggregate status for this category based on the statuses
+        of its TestResult objects."""
+        return (
+            Status.PASS
+            if all(
+                result.status == Status.PASS for result in self.test_results
+            )
+            else Status.FAIL
+        )
+
+    def add_to_workbook(self, workbook: openpyxl.Workbook) -> None:
+        """Add a sheet of failed dbt tests to an openpyxl Workbook using data
+        from the TestCategory object. Note that we expect the workbook to be
+        initialized with write_only=True."""
+        # Only add this category to the workbook if it has any failing tests
+        if self.status == Status.PASS:
+            print(
+                f"Skipping add_to_workbook for category {self.category} since "
+                f"its status is '{Status.PASS.value}'"
+            )
+            return
```

Comment:
This is the only substantive difference between this method and the previous version of the function: Now that `TestResult` and `TestCategory` store info on passing tests as well as failing tests, we need to adjust the workbook logic to skip adding `TestCategory` objects that are passing.

Reply:
Good catch, I added support for warnings in d54120e6bf2b7e2020fbfeb887fc5d7841aac83. Now the ailing rows for a warning test get logged to `test_run_result` with a `WARN` status, but they don't get added to the workbook.